plugins/plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/OpenShiftConnector.java

```diff
@@ -585,14 +594,90 @@ private boolean isTerminalAgentInjected(final Set<String> exposedPorts) {
     private boolean isDevMachine(final Set<String> exposedPorts) {
         return exposedPorts.contains(CHE_WORKSPACE_AGENT_PORT + "/tcp");
     }
-    
+
     /**
      * Che workspace id is used as OpenShift service / deployment config name
      * and must match the regex [a-z]([-a-z0-9]*[a-z0-9]) e.g. "q5iuhkwjvw1w9emg"
-     * 
+     *
      * @return randomly generated workspace id
      */
     private String generateWorkspaceID() {
         return RandomStringUtils.random(16, true, true).toLowerCase();
     }
-}
\ No newline at end of file
+
+    /**
+     * Converts a map of labels to match Kubernetes label requirements. Names are limited
+     * to alphanumeric characters, {@code '.'}, {@code '_'} and {@code '-'}, and must start and end
+     * with an alphanumeric character, i.e. they must match the regex
+     * {@code ([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9]}
+     *
+     * <p>Note that entry keys should begin with {@link OpenShiftConnector#CHE_SERVER_LABEL_PREFIX} and
+     * entries should not contain {@code '.'} or {@code '_'} before conversion;
+     * otherwise label will not be converted and included in output.
+     *
+     * @param labels Map of labels to convert
+     * @return Map of labels converted to DNS Names
+     * @see OpenShiftConnector#convertKubernetesNamesToLabels(Map)
+     */
+    protected Map<String, String> convertLabelsToKubernetesNames(Map<String, String> labels) {
+        Map<String, String> names = new HashMap<>();
+        for (Map.Entry<String, String> entry : labels.entrySet()) {
+            String key = entry.getKey();
+            String value = entry.getValue();
+
+            if (value == null) {
+                continue;
+            }
+            // Check for potential problems with conversion
+            if (!key.startsWith(CHE_OPENSHIFT_RESOURCES_PREFIX)) {
+                LOG.warn("Expected CreateContainerParams label key " + entry.getKey() +
+                         " to start with " + CHE_SERVER_LABEL_PREFIX);
+            } else if (key.contains(".") || key.contains("_") || value.contains("_")) {
+                LOG.error(String.format("Cannot convert label %s to DNS Name: contains '-' or '.'", entry.toString()));
```

Comment:
I would be more explicit in the log here: "Cannot convert label .... because these are used as escape characters."